perf: frappe.require

gavindsouza opened the pull request:
Query direct endpoint to fetch static assets (via NGINX) rather than hitting a Frappe API (via Gunicorn). This changes how the asset is queried and cached in the browser.

Fixes #22042

gavindsouza committed:
refactor: frappe.require

Query direct endpoint to fetch static assets (via NGINX) rather than
hitting a Frappe API (via Gunicorn). This changes how the asset is
queried and cached in the browser.

ankush commented:
- [ ] We should remove `localStorage` caching too. Now a useless 2nd layer of cache and it doesn't work for large assets anyway.

ankush commented on frappe/public/js/frappe/assets.js:
```diff
@@ -124,19 +124,20 @@ frappe.assets = {
+		if (items.length) frappe.dom.freeze();
+
+		items.forEach((item, idx) => {
+			fetch(item)
```
Few problems with this:

- IN prod non-bundles won't evict the cache unless we do a hard-reload after an upgrade. 
- In dev, same problem but worse because you're actively changing code and it won't refresh unless you do hard reload. 


Possible solutions:
- Send some kind of signal in boot that cache is to be avoided and hard reload is required. So EVERYONE after upgrade will get `window.location.reload(true)`
- Send no-cache header in development mode, add an exception for bundles cause bundles can be cached (invalidated by unique id)

dj12djdjs replied:
> Send some kind of signal in boot that cache is to be avoided and hard reload is required. So EVERYONE after upgrade will get window.location.reload(true)

@ankush I think this is only compatible with firefox?

I've tried this in one of my projects, I could not get it working very well.

gavindsouza replied:
Would using `frappe.boot.metadata_version` as a version string achieve the same result?

gavindsouza committed:
fix: Add version tag for non-bundled assets

ankush commented on frappe/public/js/frappe/assets.js:
```diff
@@ -124,19 +124,34 @@ frappe.assets = {
+				url.searchParams.append("v", versionString);
+			}
+			// Force use cache because we are already busting the cache with the version
+			const headers = { cache: "force-cache" };
```
Is this really required?

ankush commented on frappe/public/js/frappe/assets.js:
```diff
@@ -124,19 +124,34 @@ frappe.assets = {
+		}
+
+		frappe.dom.freeze();
+		const fetchPromises = items.map(fetchItem);
```
```suggestion
		const fetch_promises = items.map(fetch_item);
```

We use camel_case even in js :grimacing:

ankush committed:
Merge branch 'develop' into assets-fetch-no-get_js

ankush commented on frappe/public/js/frappe/assets.js:
```diff
@@ -124,19 +124,34 @@ frappe.assets = {
+
+		const versionString = frappe.boot.developer_mode
+			? Date.now()
+			: frappe.boot.metadata_version;
```
We don't need metadata version here right? window._version_number is more suitable for this.

ankush committed:
refactor: fetch logic

- Use _version_number which indicates when assets.json was generated.
  Files are evicted roughly at same time. metadata_version changes with
  other changes too.
- use `includes` instead of index check
- camelCase -> snake_case
- dont force caching. doesn't seem to be required. Caches without this
  header too.

ankush commented:
This paired with https://github.com/frappe/frappe/pull/22517 

![](https://media.tenor.com/0_-i8PEg7b4AAAAC/phir-hera-pheri-paisa-hi-paisa-hoga.gif)
